async/kafka/group/group.go

```diff
@@ -0,0 +1,179 @@
+package group
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"runtime"
+
+	"github.com/Shopify/sarama"
+	"github.com/beatlabs/patron/async"
+	"github.com/beatlabs/patron/async/kafka"
+	"github.com/beatlabs/patron/errors"
+	"github.com/beatlabs/patron/log"
+	"github.com/opentracing/opentracing-go"
+)
+
+// Factory definition of a consumer factory.
+type Factory struct {
+	name    string
+	group   string
+	topic   string
+	brokers []string
+	oo      []kafka.OptionFunc
+}
+
+// New constructor.
+func New(name, group, topic string, brokers []string, oo ...kafka.OptionFunc) (*Factory, error) {
+
+	if name == "" {
+		return nil, errors.New("name is required")
+	}
+
+	if group == "" {
+		return nil, errors.New("group is required")
+	}
+
+	if len(brokers) == 0 {
+		return nil, errors.New("provide at least one broker")
+	}
+
+	if topic == "" {
+		return nil, errors.New("topic is required")
+	}
+
+	return &Factory{name: name, group: group, topic: topic, brokers: brokers, oo: oo}, nil
+}
+
+// Create a new consumer.
+func (f *Factory) Create() (async.Consumer, error) {
+
+	config, err := kafka.SaramaConfig(f.name)
+
+	if err != nil {
+		return nil, err
+	}
+
+	cc := kafka.ConsumerConfig{
+		Brokers:      f.brokers,
+		Buffer:       0,
+		SaramaConfig: config,
+	}
+
+	c := &consumer{
+		topic:       f.topic,
+		group:       f.group,
+		traceTag:    opentracing.Tag{Key: "group", Value: f.group},
+		consumerCnf: cc,
+	}
+
+	for _, o := range f.oo {
+		err = o(c)
+		if err != nil {
+			return nil, fmt.Errorf("Could not apply OptionFunc '%v' to consumer : %v", runtime.FuncForPC(reflect.ValueOf(o).Pointer()).Name(), err)
+		}
+	}
+
+	return c, nil
+}
+
+// consumer members can be injected or overwritten with the usage of OptionFunc arguments.
+type consumer struct {
+	kafka.Consumer
+	topic       string
+	group       string
+	traceTag    opentracing.Tag
+	cnl         context.CancelFunc
+	cg          sarama.ConsumerGroup
+	consumerCnf kafka.ConsumerConfig
+}
+
+func (c *consumer) ConsumerConfig() *kafka.ConsumerConfig { return &c.consumerCnf }
```

Comment:
Any reason for this? BTW Single line functions are not common in Go.

Reply:
This is the only way I found to keep `OptionFuc` common between simple and group consumer object. 
The OptionFunc operates on `*kafka.ConsumerConfig` so it can be used with both `simple.consumer` and `group.consumer`. 

The other solution I think is to have seperate `OptionFunc`s for simple and group.

What do you think?

Reply:
Option 1: We could change the signature of OptionFunc to `type OptionFunc func(*ConsumerConfig) error`...
 Option 2: We could actually go in a totally different direction and delete all options and go as follows:
- get a DefaultConfig() *ConsumerConfig (every default value will be set up here)
- the user can change whatever he wants
- pass the whole config to simple/group consumer
Option 1 leaves the code as it is
Option 2 seems to make the code even easier...
What do you think?

Reply:
Actually If we just convert the OptionFunc to operate in `*ConsumerConfig` (option 1) then each factory just have to apply the `OptionFunc`s in `&c.config` and thats all. 

Like this
```golang
for _, o := range f.oo {
	err = o(&c.config)
	if err != nil {
		return nil, err
	}
}
```

The whole kafka package is isolated and the group/simple consumer do not need to export anything since the factory is inside these packages.

It looks pretty clear to me. And we keep the options common and isolated. 

I think we should go with that. :satisfied: What about you?

Reply:
LGTM